compiler/rustc_mir_transform/src/check_unsafety.rs

```diff
@@ -387,41 +400,109 @@ pub(crate) fn provide(providers: &mut Providers) {
-            self.unsafe_blocks.push((block.hir_id, self.used_unsafe.contains(&block.hir_id)));
+            let used = match self.tcx.lint_level_at_node(UNUSED_UNSAFE, block.hir_id) {
+                (Level::Allow, _) => Some(SomeDisallowedInUnsafeFn),
+                _ => self.used_unsafe_blocks.get(&block.hir_id).copied(),
+            };
+            // only pushes if the contained `match` doesn't `return` early
+            self.unused_unsafe.push((
+                block.hir_id,
+                match (self.context, used) {
+                    (_, None) => UnusedUnsafe::Unused,
+                    (Context::Safe, Some(_))
+                    | (Context::UnsafeFn(_), Some(SomeDisallowedInUnsafeFn)) => {
+                        let previous_context = self.context;
+                        self.context = Context::UnsafeBlock(block.hir_id);
+                        intravisit::walk_block(self, block);
+                        self.context = previous_context;
+                        return;
+                    }
+                    (Context::UnsafeFn(hir_id), Some(AllAllowedInUnsafeFn(lint_root))) => {
+                        UnusedUnsafe::InUnsafeFn(hir_id, lint_root)
+                    }
+                    (Context::UnsafeBlock(hir_id), Some(_)) => UnusedUnsafe::InUnsafeBlock(hir_id),
+                },
+            ));
+        }
+        intravisit::walk_block(self, block);
+    }
+
+    fn visit_fn(
+        &mut self,
+        fk: intravisit::FnKind<'tcx>,
+        _fd: &'tcx hir::FnDecl<'tcx>,
+        b: hir::BodyId,
+        _s: rustc_span::Span,
+        _id: HirId,
+    ) {
+        if matches!(fk, intravisit::FnKind::Closure) {
+            self.visit_body(self.tcx.hir().body(b))
         }
     }
 }
 
 fn check_unused_unsafe(
     tcx: TyCtxt<'_>,
     def_id: LocalDefId,
-    used_unsafe: &FxHashSet<hir::HirId>,
-    unsafe_blocks: &mut Vec<(hir::HirId, bool)>,
-) {
-    let body_id = tcx.hir().maybe_body_owned_by(tcx.hir().local_def_id_to_hir_id(def_id));
+    used_unsafe_blocks: &FxHashMap<HirId, UsedUnsafeBlockData>,
+) -> Vec<(HirId, UnusedUnsafe)> {
+    let hir_id = tcx.hir().local_def_id_to_hir_id(def_id);
+    let body_id = tcx.hir().maybe_body_owned_by(hir_id);
 
     let body_id = match body_id {
         Some(body) => body,
         None => {
             debug!("check_unused_unsafe({:?}) - no body found", def_id);
-            return;
+            return vec![];
         }
     };
     let body = tcx.hir().body(body_id);
-    debug!("check_unused_unsafe({:?}, body={:?}, used_unsafe={:?})", def_id, body, used_unsafe);
 
-    let mut visitor = UnusedUnsafeVisitor { used_unsafe, unsafe_blocks };
+    let context = match tcx.hir().find(hir_id) {
+        Some(Node::Item(&hir::Item { kind: hir::ItemKind::Fn(ref sig, _, _), .. }))
+            if sig.header.unsafety == hir::Unsafety::Unsafe =>
+        {
+            Context::UnsafeFn(hir_id)
+        }
+        _ => Context::Safe,
```

Comment:
Using `fn_sig_by_hir_id` as suggested above would turn the `None` case into an ICE. (I don’t know whether there even are any cases where this can fail though, anyway.